cmd/main.go

```diff
@@ -154,81 +166,123 @@ func run(kubeconfigArgs *genericclioptions.ConfigFlags, flags args, printFlags *
 		return statusFail, err
 	}
 
-	resourceCount, zombies, err := detectZombies(flags, printFlags, gitopsDynClient, clusterDynClient, clusterDiscoveryClient, gitopsRestClient, *kubeconfigArgs.Namespace)
+	resourceCount, allZombies, err := detectZombies(flags, printFlags, gitopsDynClient, clusterDynClient, clusterDiscoveryClient, gitopsRestClient, *kubeconfigArgs.Namespace)
 	if err != nil {
 		return statusFail, err
 	}
 
 	if flags.nostream {
-		err = printZombies(zombies, printFlags)
+		err = printZombies(allZombies, printFlags)
 		if err != nil {
 			return statusFail, err
 		}
 	}
 
+	var totalZombies int
+	for _, zombies := range allZombies {
+		totalZombies += len(zombies)
+	}
+
 	if flags.nostream && *printFlags.OutputFormat == "" {
-		fmt.Printf("\nSummary: %d resources found, %d zombies detected\n", resourceCount, len(zombies))
+		fmt.Printf("\nSummary: %d resources found, %d zombies detected\n", resourceCount, totalZombies)
 	}
 
-	if flags.fail && len(zombies) > 0 {
+	if flags.fail && totalZombies > 0 {
 		return statusZombiesDetected, nil
 	}
 
 	return statusOK, nil
 }
 
-func detectZombies(flags args, printFlags *k8sget.PrintFlags, gitopsDynClient, clusterDynClient dynamic.Interface, clusterDiscoveryClient *discovery.DiscoveryClient, gitopsRestClient *rest.RESTClient, namespace string) (int, []unstructured.Unstructured, error) {
+func detectZombies(flags args, printFlags *k8sget.PrintFlags, gitopsDynClient, clusterDynClient dynamic.Interface, clusterDiscoveryClient *discovery.DiscoveryClient, gitopsRestClient *rest.RESTClient, namespace string) (resourceCount int, zombies map[string][]unstructured.Unstructured, err error) {
+	zombies = make(map[string][]unstructured.Unstructured)
+
+	helmReleases, kustomizations, clustersConfigs, err := listGitopsResources(flags, gitopsDynClient, gitopsRestClient)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	ch := make(chan string)
+	var wgProducer, wgConsumer sync.WaitGroup
+
+	clustersConfigs[fluxClusterName] = clusterClients{dynamic: clusterDynClient, discovery: clusterDiscoveryClient}
+
+	wgConsumer.Add(1)
+	go func() {
+		defer wgConsumer.Done()
+		for cluster := range ch {
+			clusterResourceCount, clusterZombies, err := detectZombiesOnCluster(cluster, printFlags, flags, helmReleases, kustomizations, clustersConfigs[cluster].dynamic, clustersConfigs[cluster].discovery, namespace)
+			if err != nil {
+				klog.Errorf("[%s] could not detect zombies on: %w", cluster, err)
+			}
+
+			resourceCount += clusterResourceCount
+			zombies[cluster] = clusterZombies
+		}
+	}()
+
+	for cluster := range clustersConfigs {
+		if flags.excludeCluster != nil && slices.Contains(*flags.excludeCluster, cluster) {
+			klog.Infof("[%s] excluding from zombie detection", cluster)
+			continue
+		}
+		ch <- cluster
+	}
+
+	wgProducer.Wait()
+	close(ch)
+	wgConsumer.Wait()
```

Comment:
Clusters are scraped synchronously this way, is there any reason not handling them async? Something like
This is just some dummy code,  zombies[cluster] = clusterZombies / resourceCount +=  would open race conditions.
So it might be a nice opportunity to open the zombie channel in this function and pass it downstream.

```go
	var wg sync.WaitGroup
	clustersConfigs[fluxClusterName] = clusterClients{dynamic: clusterDynClient, discovery: clusterDiscoveryClient}

	for cluster := range clustersConfigs {
		if flags.excludeCluster != nil && slices.Contains(*flags.excludeCluster, cluster) {
			klog.Infof("[%s] excluding from zombie detection", cluster)
			continue
		}

                wg.Add(1)
	        go func(cluster string) {
		        defer wg.Done()
			        clusterResourceCount, clusterZombies, err := detectZombiesOnCluster(cluster, printFlags, flags, helmReleases, kustomizations, clustersConfigs[cluster].dynamic, clustersConfigs[cluster].discovery, namespace)
			        if err != nil {
				        klog.Errorf("[%s] could not detect zombies on: %w", cluster, err)
			        }
        
			        resourceCount += clusterResourceCount
			        zombies[cluster] = clusterZombies
	        }(cluster)
        }
	wg.Wait()

```

Reply:
Arf ... I don't know what I've done, sorry 🙏 It should be better this way 😜